File: 141-linked-list-cycle/linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_map<ListNode*,int>nodesCnt;

        ListNode* temp = head;

        while(temp!=NULL){
            nodesCnt[temp]++;

            if(nodesCnt[temp]>1){
                return true;
            }
            temp = temp->next;
        }
        
        return false;
    }
};

File: 141-linked-list-cycle/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fst = head;

        while(fst!=null && fst.next!=null)
        {
            fst = fst.next.next;
            slow = slow.next;

            if(fst==slow){
                return true;
            }
        }
        return false;
    }
}